tests/NSubstitute.Acceptance.Specs/ArgumentMatching.cs

```diff
@@ -740,9 +740,85 @@ public void Supports_custom_argument_matcher_descriptions()
         Assert.That(ex.Message, Contains.Substring("24 is not forty two"));
     }
 
+    [Test]
+    public void Supports_matching_generic_interface_bound_type_string_with_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MyStringArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<IMyArgument<string>>());
+    }
+
+    [Test]
+    public void Supports_matching_generic_interface_bound_type_custom_class_with_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MySampleClassArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<IMyArgument<SampleClass>>());
+    }
+
+    [Test]
+    public void Supports_matching_generic_interface_bound_type_custom_class_with_derived_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MySampleDerivedClassArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<IMyArgument<SampleClass>>());
+    }
+
+    [Test]
+    public void Supports_matching_custom_class_with_derived_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MySampleDerivedClassArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<MySampleClassArgument>());
+    }
+
+    [Test]
+    public void Supports_matching_generic_interface_bound_type_ArgAnyType_with_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MyStringArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<IMyArgument<Arg.AnyType>>());
+    }
+
+    [Test]
+    public void Supports_matching_generic_interface_bound_type_ArgAnyType_with_derived_class_argument()
+    {
+        var service = Substitute.For<IMyService>();
+        var argument = new MySampleDerivedClassArgument();
+
+        service.MyMethod(argument);
+
+        service.Received().MyMethod(Arg.Any<IMyArgument<Arg.AnyType>>());
```

Comment:
**question:** does `DidNotReceive()` also work as expected with these changes? e.g. `Any<MyStringArgument>` should not match a different `IMyArgument<string>` arg passed. Could you please add a few tests around that too?

Reply:
Thank you for the update, @dtchepak!

I have added two tests covering this constellation, and they work correctly with the implementation in the PR.